### ISSUE TITLE
[Memory] Issues with Game Memory

### ISSUE THREAD

**Issue by nlgzrgn** [ExOptsTeam/NFSMWExOpts#4]
- [x] Car models flicker when there are lots of objects (cars, etc.) (Especially when it rains)
[Imgur](http://i.imgur.com/PTNMMon.jpg)
[Imgur](http://i.imgur.com/0nFXK5E.jpg)
Also check out the screenshots in 5th comment.

- [ ] Some opponents have ghosted cars if you choose more than 3 opponents in Quick Race mode.
Note 1: This happens even if the cars are totally the same.
Note 2: It will less likely happen when opponents or you choose police/traffic cars.
Note 3: This also limits their count. If every opponent take a race car, max. 6 will spawn.

**Comment by Aero52** [comment]
Is it possible this could be caused by the increased RVM LOD that was added to the widescreen fix? Perhaps it's too much geometry at once for the game engine.

**Comment by nlgzrgn** [comment]
Thanks for the info.

Does disabling RearviewMirrorFix from Widescreen fix also disables geometry detail for it?

**Comment by Aero52** [comment]
Apparently so, yes.

And if you want to test this in real-time, all that was done to increase the RVM LOD was this:

```
// change jne to jmp
004FAEB0 jmp 004FAEB5 

// nop je 006BFE57
006BFE33 nop 
006BFE34 nop
```

**Comment by Aero52** [comment]
Forgot to mention that I discovered something that restores missing geometry. Maybe it solves the problem with vanishing cars?

008FAE44 00000000

![Geometry](http://i.imgur.com/1EItxng.gif)

**Comment by nlgzrgn** [comment]
Thank you again for another fix! But that's not the thing we want. Disappearing cars issue still persists.

How to reproduce it:
- Get in a x5 or x6 pursuit. There will be lots of cars.
- Make it rain. (Wait for it to rain randomly or force it from ExOpts) It happens especially when rain is enabled.

I have some pics below. I was surrounded there. Cars were flickering:

![nfsmw065](https://user-images.githubusercontent.com/1937572/27332390-60f239d8-55ca-11e7-9122-de9c8659b7d0.jpg)
![nfsmw066](https://user-images.githubusercontent.com/1937572/27332389-60ef0722-55ca-11e7-94be-e46ec29e0fa6.jpg)

Also changed the title.

**Comment by DiamondMaster07** [comment]
High-Res shadows(4096 for ex.) can really help out reproducing this glitch and making it worse

**Comment by Aero52** [comment]
@nlgzrgn Thanks for the information. I'll take a look later.

@DiamondMaster07 It might just be a coincidence, but it wouldn't surprise me since shadows need a lot of work still. Based on my research, 4096 shouldn't even be necessary at 1080p, but the game doesn't render shadows at full quality, so it needs a much higher value (more memory) than it should.

**Comment by Aero52** [comment]
While I couldn't fix anything, I've confirmed that the rearview mirror makes the problem worse. It's significant enough that disabling it can cause all of the cars to reappear again. I'll ask @ThirteenAG to remove [this code](https://github.com/ThirteenAG/WidescreenFixesPack/blob/master/source/NFSMostWanted.WidescreenFix/dllmain.cpp#L330-L331) to partially reduce the geometry in the RVM, since some of it looks terrible anyway.

ThirteenAG, you can also add this to restore [missing geometry](https://camo.githubusercontent.com/46a2d8b8959bf98b1bc88995b4ad2f0fa3588076/687474703a2f2f692e696d6775722e636f6d2f31454974786e672e676966) and leave it always on if you prefer.

> 006C69A7 mov al,00 // restores missing geometry

@nlgzrgn You probably know this already, but I'm pretty sure the rain makes it worse because of the geometry in the road reflections.

@DiamondMaster07 I couldn't confirm shadow resolution had any affect, but while I was testing I found something that fixes the problem with shadows randomly disappearing at 006C9653. The only thing left to do is increase the render size to fix the ["shadow tearing"](https://www.youtube.com/watch?v=nvdfcmn3Tek) issue, but I haven't had any luck with that so far.

**Comment by nlgzrgn** [comment]
I think that will be better for now, too. Thank you for your findings again!
Btw, if you have a discord, please hit me there: nlgzrgn#3258 ;)

**Comment by berkay2578** [comment]
I'm not able to reproduce the bug. Proof: https://gfycat.com/WideeyedLimpingLamprey

**Comment by DiamondMaster07** [comment]
@berkay2578 Maybe you don't have ws fix on

**Comment by berkay2578** [comment]
@DiamondMaster07 oh ok, I read the issue as if the problem was with the game itself (since it's titled `Issues with Game Memory`).

**Comment by AeroWidescreen** [comment]
It's a combination of road reflections (rain) + rear view mirror + traffic + police. We're guessing it's too much geometry at once. The rear view mirror has a significant impact, so if that isn't visible then you might never notice a problem.

And just so you know, you can use the widescreen fix even if you don't (or can't) play in widescreen.

**Comment by berkay2578** [comment]
@AeroWidescreen ws fix + rain + traffic + police: https://gfycat.com/EquatorialScaryArmyant I'll try again with the rear view mirror.

**Comment by DiamondMaster07** [comment]
@berkay2578 
Here's without shadow fix: https://gfycat.com/JoyfulZigzagCentipede
Models still glitch around also the rearview mirror is bugged as hell.
I have my RoadReflectionAmmount on 3 in the ExOpts ini file

**Comment by berkay2578** [comment]
Since I'm unable to reproduce this issue _at all_, here's my environment:
PC:
 - Intel i5 3337u
 - Nvidia GT 740m
 - 8 GB ram, no virtual ram
 - sata3 ssd
Game:
 - v1.3 with no file mods
 - [4GB address patched](http://www.ntcore.com/4gb_patch.php)
 - [ExtOpts](https://hastebin.com/atiqabiyok)
 - [WS Fix](https://hastebin.com/onabaqofoc)
 - RockportEd (the thing you see in my gifs), forces `D3DRS_CULLMODE` to `D3DCULL_NONE` but I don't think that will have any effect on this.

**Comment by nlgzrgn** [comment]
I've tested it again with and without ShadowsFix (also with some settings similar to Berkay's), but I still have that issue.
![nfsmw088](https://user-images.githubusercontent.com/1937572/29623595-9222c8be-882f-11e7-9f9d-5dc74dca6343.jpg)
![nfsmw087](https://user-images.githubusercontent.com/1937572/29623594-91a9a7f4-882f-11e7-9449-b386e88903fe.jpg)

Something else?

**Comment by DiamondMaster07** [comment]
@berkay2578 Try editing RoadReflectionAmmount in ExOpts ini to higher(too high values make the road too bright), also don't smack up the cars, let them go.
Here's what I did:
1. Set my heat level to 5+x
2. Started the 2nd bounty challenge(Rosewood fire station) with rain forced and RoadRefl.Ammo. at 3
3. I went instantly to the bus station to make the police glitch
4. Waited for some seconds

**Comment by berkay2578** [comment]
I already have it at 3, you can check it in the link I posted. Also, I stack the cars to ensure it's not directx related. Regardless, if it _is_ because of the reflections, that only means the EAGL inside the game never was made to draw meaningful reflections. A way around it would be to hook the limits and remove them (since we don't run 2005 hardware anymore), GTA:SA has a similar thing called 'Limit Adjuster'. Attempting to do something like that would take considerable amount of time and effort. 

I'll try to port my LOD Controller to Most Wanted, see if the game receives NULL on load (d3d9 limit, maybe forcing d3d9ex will help. I think ReShade tries to do this) or if the game fails on load (low-level software limit, basically EAGL limit). I'm not sure about any of these to be honest, I've only been messing around with Most Wanted for maybe a week and I also can't reproduce the issue.

**Comment by AeroWidescreen** [comment]
@berkay2578  Wow, I wasn't expecting to read this today... amazing breakthrough! Thanks for all the information, I look forward to seeing what solution you come up with. I knew something was odd, since you weren't able to reproduce the problem at all.

**Comment by DiamondMaster07** [comment]
I don't know if this can go here, but it is somehow possible to make the rearview mirror render properly?
You can see all the things behind me:
![need for speed underground speed exe 08 24 2017 - 19 45 41 03](https://user-images.githubusercontent.com/26389402/29680529-20569ee6-8905-11e7-9664-b1ad0095e756.png)
In the rearview mirror, they don't even render
![need for speed underground speed exe 08 24 2017 - 19 45 37 02](https://user-images.githubusercontent.com/26389402/29680636-7ab1a944-8905-11e7-9250-fc7d6da7f014.png)

Also it has serious issues with world rendering
What's behind me
![need for speed underground speed exe 08 24 2017 - 19 55 52 04](https://user-images.githubusercontent.com/26389402/29680901-530bd22e-8906-11e7-86ba-b992c3496ba8.png)
What the mirror renders
![need for speed underground speed exe 08 24 2017 - 19 55 56 05](https://user-images.githubusercontent.com/26389402/29680977-969a63a2-8906-11e7-86ac-7c2dff8b631a.png)
(Don't mind the shadow bug above)

Also, it does not render any particles
Police lights for example
![need for speed underground speed exe 08 24 2017 - 20 04 54 09](https://user-images.githubusercontent.com/26389402/29681608-87bc8d54-8908-11e7-8d43-8ea9c74064ba.png)
And they don't render in the mirror(and again, you can see the poor world rendering)
![need for speed underground speed exe 08 24 2017 - 20 05 02 10](https://user-images.githubusercontent.com/26389402/29681629-984417fa-8908-11e7-8a91-b6b5183de0cb.png)

Note: if you think this post should go somewhere else, then tell me

**Comment by berkay2578** [comment]
I didn't have that issue in, like, ever. Try with a complete clean installation of NFS: Most Wanted and troubleshoot your way in to find the culprit.

**Comment by nlgzrgn** [comment]
Alright, it looks lile Berkay found the 1st thing! (Btw, @berkay2578 and @AeroWidescreen, hit me on Discord asap.) (nlgzrgn#3258)

Now we focus on the 2nd one. A test build with this fix will be available soon for our testers.

**Comment by berkay2578** [comment]
Do you mean like these? http://i.imgur.com/zGC7qbF.png

**Comment by DiamondMaster07** [comment]
I reinstalled the game but I still have that bug.
![need for speed underground speed exe 08 24 2017 - 22 03 23 13](https://user-images.githubusercontent.com/26389402/29686124-24089d10-8918-11e7-84d2-0aca9a14b742.png)
Don't mind the misplaced hud

**Comment by AeroWidescreen** [comment]
> I don't know if this can go here, but it is somehow possible to make the rearview mirror render properly?
> You can see all the things behind me:

We already did this in the widescreen fix, but then partially downgraded it in the last update because of this problem. The amount detail you currently see is about the same as Carbon. We might add the other part back, but those extra details were kind of ugly due to the weird lighting in my opinion.

> Alright, it looks lile Berkay found the 1st thing! (Btw, @berkay2578 and @AeroWidescreen, hit me on Discord asap.) (nlgzrgn#3258)

I don't normally use Discord, but I'll get on it in a few hours if you need help.

**Comment by nlgzrgn** [comment]
Update for 2nd issue here:

> Some opponents have ghosted cars if you choose more than 3 opponents in Quick Race mode.
We added an experimental skin fix into the new ExOpts, v7.0.1.1337.
The fix simply tries to use 2nd, 3rd and 4th players' skins in the next vehicles as well.
That issue here was with DUMMY_SKIN_X and DUMMY_WHEEL_X textures in CARS\TEXTURES.BIN. (X is any number)

> This also limits their count. If every opponent take a race car, max. 6 will spawn.
Limit thing is something else. It's totally not related with skin issue. I was able to spawn more cars by editing this value:
`.data:008F7EF0 CarLoaderPoolSizes dd 50000 `
But the problem is, increasing that value messes with the world renderer.
